application/src/main/java/org/thingsboard/server/service/install/update/DefaultDataUpdateService.java

```diff
@@ -203,11 +219,100 @@ public void updateData(String fromVersion) throws Exception {
+                                clazzName,
+                                fromVersion,
+                                currentVersion);
+                        try {
+                            TbPair<Boolean, JsonNode> upgradeRuleNodeConfigurationResult = versionedNode.upgrade(ruleNodeId, fromVersion, oldConfiguration);
+                            if (upgradeRuleNodeConfigurationResult.getFirst()) {
+                                ruleNode.setConfiguration(upgradeRuleNodeConfigurationResult.getSecond());
+                                var ruleChainId = ruleNode.getRuleChainId();
+                                var tenantId = getTenantId(ruleChainIdToTenantIdMap, ruleNodeId, ruleChainId);
+                                if (tenantId == null) {
+                                    log.warn("Failed to find tenant id for rule chain with id: {}", ruleChainId);
+                                    continue;
+                                }
+                                ruleChainService.saveRuleNode(tenantId, ruleNode);
+                                log.info("Successfully upgrade rule node with id: {} type: {} fromVersion: {} toVersion: {}",
+                                        ruleNodeId,
+                                        clazzName,
+                                        fromVersion,
+                                        currentVersion);
+                            }
+                        } catch (TbNodeException e) {
+                            log.warn("Failed to upgrade rule node with id: {} type: {} fromVersion: {} toVersion: {} due to: ",
+                                    ruleNodeId,
+                                    clazzName,
+                                    fromVersion,
+                                    currentVersion,
+                                    e);
+                        }
+                    }
+                }
+            }
+            log.info("Finished rule nodes upgrade!");
+        } catch (Exception e) {
+            log.error("Unexpected error during rule nodes upgrade: ", e);
+        }
+    }
+
+    private TenantId getTenantId(HashMap<RuleChainId, TenantId> ruleChainIdToTenantId, RuleNodeId ruleNodeId, RuleChainId ruleChainId) {
+        return ruleChainIdToTenantId.computeIfAbsent(ruleChainId,
+                id -> {
+                    RuleChain ruleChain = ruleChainService.findRuleChainById(TenantId.SYS_TENANT_ID, id);
+                    if (ruleChain == null) {
+                        log.warn("Failed to find rule chain by id: {} ruleNodeId: {}", ruleChainId, ruleNodeId);
+                        return null;
+                    }
+                    return ruleChain.getTenantId();
+                });
+    }
+
+    private Set<BeanDefinition> getBeanDefinitions(Class<? extends Annotation> componentType) {
+        ClassPathScanningCandidateComponentProvider scanner = new ClassPathScanningCandidateComponentProvider(false);
```

Comment:
Please reuse AnnotationComponentDiscoveryService or, maybe better - extract this logic to a separate service.

![image](https://github.com/thingsboard/thingsboard/assets/5016373/41f48393-6187-4869-aa7a-ba26c3f8fffb)